consensus/istanbul/proxy/validator_assignment.go

```diff
@@ -0,0 +1,233 @@
+//           the user's responsibility to ensure that.
+type consistentHashingPolicy struct {
+	c      *consistent.Consistent // used for consistent hashing
+	logger log.Logger
+}
+
+func newConsistentHashingPolicy() *consistentHashingPolicy {
+	// This sets up a consistent hasher with bounded loads:
+	// https://ai.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html
+	// Partitions are assigned to members (proxies in this case).
+	// using a hash ring.
+	// When locating a validator's proxy using `LocateKey`, the validator is assigned
+	// to a partition using hash(validator's address) % PartitionCount in constant time.
+	cfg := consistent.Config{
+		// Prime to distribute validators more uniformly.
+		// Higher partition count generally gives a more even distribution
+		PartitionCount: 271,
+		// The number of replications of a proxy on the hash ring
+		ReplicationFactor: 40,
+		// Used to enforce a max # of partitions assigned per proxy, which is
+		// (PartitionCount / len(proxies)) * Load. A load closer to 1 gives
+		// more uniformity in the # of partitions assigned to specific members,
+		// but a higher load results in less relocations when proxies are added/removed
+		Load:   1.2,
+		Hasher: hasher{},
+	}
+
+	return &consistentHashingPolicy{
+		c:      consistent.New(nil, cfg),
+		logger: log.New(),
+	}
+}
+
+// assignProxy adds a proxy to the consistent hasher and recalculates all validator assignments
+func (ch *consistentHashingPolicy) assignProxy(proxy *Proxy, valAssignments *valAssignments) bool {
+	ch.c.Add(proxy.ID())
+	return ch.reassignValidators(valAssignments)
+}
+
+// removeProxy removes a proxy from the consistent hasher and recalculates all validator assignments
+func (ch *consistentHashingPolicy) removeProxy(proxy *Proxy, valAssignments *valAssignments) bool {
+	ch.c.Remove(proxy.ID().String())
+	return ch.reassignValidators(valAssignments)
+}
+
+// assignRemoteValidators adds remote validators to the valAssignments struct and recalculates
+// all validator assignments
+func (ch *consistentHashingPolicy) assignRemoteValidators(vals []common.Address, valAssignments *valAssignments) bool {
+	valAssignments.addValidators(vals)
+	return ch.reassignValidators(valAssignments)
+}
+
+// removeRemoteValidators removes remote validators from the valAssignments struct and recalculates
+// all validator assignments
+func (ch *consistentHashingPolicy) removeRemoteValidators(vals []common.Address, valAssignments *valAssignments) bool {
+	valAssignments.removeValidators(vals)
+	return ch.reassignValidators(valAssignments)
+}
+
+// reassignValidators recalculates all validator <-> proxy pairings
+func (ch *consistentHashingPolicy) reassignValidators(valAssignments *valAssignments) bool {
+	logger := ch.logger.New("func", "reassignValidators")
+	anyAssignmentsChanged := false
+	for val, proxyID := range valAssignments.valToProxy {
+		newProxyID := ch.c.LocateKey(val.Bytes())
+
+		valReassigned := false
```

Comment:
I believe valReassigned is redundant.  We could directly set `anyAssignmentsChanged` to true when a change is made, and then the `if !anyAssignmentsChanged && valReassigned {` block would also go away.

Reply:
Yes, that makes sense.  Made the change.